src/lexer/lexer.go

```diff
@@ -0,0 +1,482 @@
+					}
+				}
+			} else if failure, ok := sToks.tokens.(Failure); ok {
+				return nil, failure.Error(lineno, filename)
+			} else if toks, ok := sToks.tokens.(string); ok {
+				// strings as token just are yielded as it.
+				data := grp
+				// update brace / parentheses balance
+				if toks == TokenOperator {
+					switch data {
+					case "{":
+						balancingStack.Push("}")
+					case "(":
+						balancingStack.Push(")")
+					case "[":
+						balancingStack.Push("]")
+					case "}", ")", "]":
+						exOp := balancingStack.Pop()
+						if exOp == nil {
+							return nil, errors.TemplateSyntaxError(fmt.Sprintf("unexpected '%s'", data), lineno, name, filename)
+						}
+						if *exOp != data {
+							return nil, errors.TemplateSyntaxError(fmt.Sprintf("unexpected '%s', expected '%s'", data, *exOp), lineno, name, filename)
+						}
+					}
+				}
+
+				// yield items
+				if data != "" || !ignoreIfEmpty.Has(toks) {
+					ret = append(ret, tokenRaw{lineno, toks, data})
+				}
+				lineno += strings.Count(data, "\n")
+			} else {
+				return nil, fmt.Errorf("unexpected type")
+			}
+
+			lineStarting = grp[len(grp)-1] == '\n'
+			// fetch new position into new variable so that we can check
+			// if there is a internal parsing error which would result
+			// in an infinite loop
+			idx := sToks.pattern.FindAllStringSubmatchIndex(source[pos:], -1)
+			pos2 := pos + idx[0][1]
+			// handle state changes
+			if sToks.command != nil {
+				// remove the uppermost state
+				if *sToks.command == "#pop" {
+					st.Pop()
+				} else if *sToks.command == "#bygroup" {
+					// resolve the new state by group checking
+					names := sToks.pattern.SubexpNames()[1:]
+					found := false
+					for i := 0; i < len(names); i++ {
+						if names[i] != "" && groups[i] != "" {
+							st.Push(names[i])
+							found = true
+							break
+						}
+					}
+					if !found {
+						return nil, fmt.Errorf("'%s' wanted to resolve the new state dynamically but no group matched", sToks.pattern)
+					}
+				} else {
+					st.Push(*sToks.command)
+				}
+
+				stateTokens = l.Rules[*st.Peek()]
+			} else if pos2 == pos {
+				// we are still at the same position and no stack change.
+				// this means a loop without break condition, avoid that and
+				// raise error
+				return nil, fmt.Errorf("'%s' yielded empty string without stack change", sToks.pattern)
+			}
+			// publish new function and start again
+			pos = pos2
+			broke = true
+			break
+		}
+	}
+
+	// if loop terminated without break we haven't found a single match
+	// either we are at the end of the file or we have a problem
+	if pos >= sourceLength {
+		return
+	}
+	return nil, errors.TemplateSyntaxError(fmt.Sprintf("unexpected char '%s' at %d", string(source[pos]), pos), lineno, name, filename)
+}
+
+type Failure struct {
+	msg string
+}
+
+func (f Failure) Error(lineno int, filename *string) error {
+	// I do not undestand why filename is passed as name and not filename but that what jinja does.
```

Comment:
Maybe that's a mistake in jinja?

Reply:
probably though I am not sure if I should leave it as it is or fix it. What do you think?

Reply:
Yeah, we should leave it as it is, even if it's a mistake it doesn't hurt anybody.